src/main/java/org/truffleruby/parser/BodyTranslator.java

```diff
@@ -3076,7 +3077,14 @@ public RubyNode visitReturnNode(ReturnParseNode node) {
 
         RubyNode translatedChild = translateNodeOrNil(sourceSection, node.getValueNode());
 
-        final RubyNode ret = new ReturnNode(environment.getReturnID(), translatedChild);
+        final RubyNode ret;
+
+        if (environment.isBlock()) {
```

Comment:
Could we also use `LocalReturnNode` if we are inside a syntactic lambda (`-> { return 42 }`)?

Reply:
Do you know off the top of your head how to find out we're inside a syntactic lambda? I can't find the method in the environment.

Reply:
I'd suggest such a change goes into its own PR. I don't think we record enough information in the environment to do it yet, and it can only ever be done for `->` style lambdas unless we can rewrite the parse tree if the assumption that `lambda` hasn't been redefined is invalidated.

Reply:
OK, let's do that as a separate PR. Indeed, just for `-> {}` lambdas